File: internal/domain/snapshot.go
package domain

import (
	"time"
)

type Snapshot struct {
	SnapshotID string                 `json:"snapshot_id" gorm:"primaryKey;type:varchar(64)"`
	Timestamp  time.Time              `json:"timestamp" gorm:"index"`
	Metrics    map[string]float64     `json:"metrics" gorm:"type:jsonb"`
	Dimensions map[string]string      `json:"dimensions" gorm:"type:jsonb"`
	CreatedAt  time.Time              `json:"created_at"`
}

func (Snapshot) TableName() {

}

func (Snapshot) TableName() string {
	return "snapshots"
}

File: trace-topology-domain/src/main/java/com/tracetopology/domain/entity/Snapshot.java
package com.tracetopology.domain.entity;

import com.tracetopology.common.utils.IdGenerator;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.time.Instant;
import java.util.Map;

@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Snapshot {

    private String snapshotId;
    private Instant timestamp;
    private Map<String, Double> metrics;
    private Map<String, String> dimensions;

    public static Snapshot create(Map<String, Double> metrics, Map<String, String> dimensions) {
        return Snapshot.builder()
                .snapshotId(IdGenerator.generateId("snap"))
                .timestamp(Instant.now())
                .metrics(metrics)
                .dimensions(dimensions)
                .build();
    }

    public Double getMetric(String name) {
        return metrics.get(name);
    }

    public String getDimension(String name) {
        return dimensions.get(name);
    }
}
